### PR TITLE
Allow specifying TLS options with internode_encryption=none + add "transitional" mode

### LINKED ISSUE #18903: Seamless transition to inter-node encryption
Currently, transitioning from `internode_encryption="none"` to any other mode (`all`, `dc`, or `,rack`)
is painful, since the procedure requires rolling restart and nodes that would like to start encrypting their internode communication need their peers to listen on the ssl port already.

To achieve that, the proposal is to always listen on `ssl_storage_port` is it's not set to 0, and the `certificate` and `keyfile` attributes are provided in `server_encryption_options` regardless of the `internode_encryption` mode.

With that, the procedure to enable internode encryption would be:
1. Set `internode_encryption="none"`, `certificate=<path to cert file>`, and `keyfile=<path to key file>` in scylla.yaml
2. Do cluster rolling restart (after which all nodes will listen on the `ssl_storage_port`, but will still use the unencrypted `storage_port` to send RPCs)
3. Change `internode_encryption` to one of `all`, `dc`, or `rack`
4. Do cluster rolling restart (after which all nodes will transition to the desired internode_encryption mode, will listen on the ssl_storage_port, and will listen also on `storage_port` by default for `dc` and `rack` modes, and will use the encrypted port depending on the mode and the peer node network topology locality.

Documentation also needs to be updated to reflect that procedure.

### PR DESCRIPTION
**Pull request by elcallio**
Fixes #18903
    
Adds a "transitional" internode encryption mode, under which all _outgoing_ RPC connections will use TLS, but we will still accept any incoming non-tls connection.

This allows an operator to perform a move to TLS RPC without cluster downtime:

1. For each server, add certificate etc options to server_encryption_options + internode_encryption=none + set ssl_storage_port + restart (rolling)

2. For each server, set internode_encryption=transitional + RR
3. For each server, set internode_encryption=all + RR

### CODE REVIEW

File: message/messaging_service.hh
Comment (bhalevy) [inline message/messaging_service.hh:263]: I don't see support for it in scylla_main

**Comment by bhalevy** [comment]
Does the transitional option also allow transitioning from `all` back to `none` (or `dc` or `rack`)?
It seems so, since the change from `all` to `transitional` will listen on the unecrypted port (assuming it's set to a non-zero value).
Then we switch to `none` in RR, and when that round is done,
we can disable the ssl_storage_port, if we wish to.

1. Is it simpler/safer (i.e. less susceptible to human error, and easier to document/explain to admins) than just making listening on storage_port and ssl_storage_port independent of the encryption mode?
2. Are the above config options live-updateable so the admin doesn't have to rolling-restart all nodes, just do rolling signal to reload config on all nodes?

**Comment by nyh** [comment]
> Adds a "transitional" internode encryption mode, under which all _outgoing_ RPC connections will use TLS, but we will still accept any incoming non-tls connection.

Is this the right direction? If a node comes up with using TLS for outgoing connections, it will not be able to connect to nodes which still haven't made this transition. Shouldn't it be the other way around - i.e., the nodes should start listening to TLS connections but not create any until all the nodes have switched?

I'm sure you understand this issue better than me so I must be missing something, so can you please tell me where I'm misunderstanding?

File: message/messaging_service.hh
Comment (elcallio) [reply message/messaging_service.hh:263]: Sorry. Managed to fail adding the file.

**Comment by elcallio** [comment]
> Is this the right direction? If a node comes up with using TLS for outgoing connections, it will not be able to connect to nodes which still haven't made this transition. Shouldn't it be the other way around - i.e., the nodes should start listening to TLS connections but not create any until all the nodes have switched?
> 
> I'm sure you understand this issue better than me so I must be missing something, so can you please tell me where I'm misunderstanding?

Yes. This is why the first step, enabling TLS but having `internode_encryption=none` is required first. This will ensure all servers _have_ a TLS connector and will accept both TLS and non-TLS indiscriminately, but will not initiate TLS connections. 

The second step, we enable ´transitional`, thus each node will start _creating_ TLS connections, but not require incoming to be TLS. 

The third step, we set mode to `all` (or optionally dc or rack, but again, should be considered deprecated), which will cause us to both initiate TLS and reject non-TLS (unless dc/rack conditions match).

**Comment by nyh** [comment]
> > Is this the right direction? If a node comes up with using TLS for outgoing connections, it will not be able to connect to nodes which still haven't made this transition. Shouldn't it be the other way around - i.e., the nodes should start listening to TLS connections but not create any until all the nodes have switched?
> > I'm sure you understand this issue better than me so I must be missing something, so can you please tell me where I'm misunderstanding?
> 
> Yes. This is why the first step, enabling TLS but having `internode_encryption=none` is required first. This will ensure all servers _have_ a TLS connector and will accept both TLS and non-TLS indiscriminately, but will not initiate TLS connections.

Oh, I didn't understands that. So "none" doesn't really mean "none"? :-)

Maybe you can explain this in the commit message, but now that I think of this: How will any user figure out this complicated process? Are they expected to read about it in the git log? I think your patch should include some documentation that will mention this process. Either stand-alone documentation, or an improvement to some existing documentation of internode_encryption, or maybe even just an comment inside scylla.yaml saying what internode_encription=transitional mean. But I think it should be documented in one of these ways, or nobody will ever figure out how to use it correctly.

**Comment by elcallio** [comment]
It is explained in the first patch message. 
Yes, I will add docs, but before I do I want a blessing or rejection from the peanut gallery. :-)

**Review by bhalevy — APPROVED** [review]
LGTM.

Should we open another issue for live updates of the relevant params?
(assuming sending SIGHUP to reload the config when multiple config options are need to change together.  With the config virtual table we don't have such an "apply button" so the order of changes matters, maybe unless they are done in a CQL batch? 
 not sure it will help)

**Comment by elcallio** [comment]
I think so, yes. Doing live update of network setup has a lot of pitfalls and potential races. Also, it would be  quite a big refactoring.

**Comment by bhalevy** [comment]
> I think so, yes. Doing live update of network setup has a lot of pitfalls and potential races. Also, it would be quite a big refactoring.

https://github.com/scylladb/scylladb/issues/18965

**Comment by xemul** [comment]
Dequeued, it seems to clashed with scylladb/scylla-dtest@dd3e1a0044daa4f9fafe22d61c7be7ca0f6e886c which propagated to its master first

**Comment by xemul** [comment]
Also https://github.com/scylladb/scylla-dtest/issues/4211

**Comment by bhalevy** [comment]
This series (2c189a516011446123553535614318cd1310a9f0 in particular) conflicts with https://github.com/scylladb/scylla-dtest/pull/4342
apparently since it allows listening on the `ssl_storage_port` in `none` internode_encryption` mode.

@scylladb/scylla-maint please dequeue

**Comment by xemul** [comment]
> @scylladb/scylla-maint please dequeue

EALREADY

**Comment by elcallio** [comment]
I would suggest waiting also until a test to _verify_ this works as intended (and no other issues turn up) before merging.

**Comment by bhalevy** [comment]
> I would suggest waiting also until a test to _verify_ this works as intended (and no other issues turn up) before merging.

Yes.  That would be scylladb/scylla-dtest#4372
That we could use to verify this PR before and while it's promoted.

**Comment by elcallio** [comment]
See https://github.com/scylladb/scylla-dtest/pull/4395. 
Not sure how to make a unit test. If you mean boost test, this does neither support clustering or really relevant scylla options. 
If you mean any of the python frameworks under test, then maybe... I am not familiar with how to do clustered + stress there, or if it is even possible...

**Comment by bhalevy** [comment]
> See [scylladb/scylla-dtest#4395](https://github.com/scylladb/scylla-dtest/pull/4395). Not sure how to make a unit test. If you mean boost test, this does neither support clustering or really relevant scylla options. If you mean any of the python frameworks under test, then maybe... I am not familiar with how to do clustered + stress there, or if it is even possible...

I meant a single node test that would just verify that scylla listens on ports as expected based on the configuration (or lack of it) of storage_port and ssl_storage_port in each of the supported encryption modes. (I.e pretty much of what the current dtest does, just as a unit test so it can be modified along with any changes in this area)

**Comment by elcallio** [comment]
My only issue with that is that we have no good way to verify messaging server listerning (or rather _accepting_ on given ports). The current test checks logs, which is fine in the sense that it verifies the port is open. But part of the internode thing is not just open/closed port, it is also accepting incoming connections, which is the big difference between "transitional" and something like "dc" or "rack".

**Comment by bhalevy** [comment]
I'm ok with dtest-level cluster test as well.
How would you verify that the unencrypted port is accessible only within a dc by the way? (in "dc" encryption mode)

**Comment by elcallio** [comment]
The point is that while both TLS and non-TLS ports are open in "dc" mode, any connection from a non-same-dc host are actively refused _after_ accept. Technically one should verify that this is not the case in transitional mode. But again, we have no good way to check whether rpc is accepted beyond running nodes against each other.

**Comment by bhalevy** [comment]
> The point is that while both TLS and non-TLS ports are open in "dc" mode, any connection from a non-same-dc host are actively refused _after_ accept. Technically one should verify that this is not the case in transitional mode. But again, we have no good way to check whether rpc is accepted beyond running nodes against each other.

Understood.
I wonder if dc/rack modes are even in use.
They don't make much sense beyond squeezing the last drop of cpu performance from intra-dc / rack communications.

**Comment by bhalevy** [comment]
@elcallio we still have a practical issue that the existing dtest, which is gating, fails with this change, since it is tailored to the existing semantics.
That's the reason I wanted to have an equivalent unit test to replace it.
We can take the dtest out of gating in advance so we can merge this PR and then drop it, assuming we include a unit test here to replace the dtest.
Maintaining this low-level dtest there is a hassle.

**Comment by mykaul** [comment]
@scylladb/scylla-maint - please review.

**Comment by denesb** [comment]
> @scylladb/scylla-maint - please review.

The PR is reviewed and approved, but there is an outstanding action item blocking the merge: https://github.com/scylladb/scylladb/pull/18939#issuecomment-2148811339.

**Comment by elcallio** [comment]
> > @scylladb/scylla-maint - please review.
> 
> The PR is reviewed and approved, but there is an outstanding action item blocking the merge: [#18939 (comment)](https://github.com/scylladb/scylladb/pull/18939#issuecomment-2148811339).

See the push beneath the comment: Test requested was added.

**Comment by denesb** [comment]
> > > @scylladb/scylla-maint - please review.
> > 
> > 
> > The PR is reviewed and approved, but there is an outstanding action item blocking the merge: [#18939 (comment)](https://github.com/scylladb/scylladb/pull/18939#issuecomment-2148811339).
> 
> See the push beneath the comment: Test requested was added.

So the requested test is in? Or we just have an open PR? In the latter case, please link it here, so we can monitor it.

**Comment by elcallio** [comment]
PR (not committed, mainly because it requires this patch): https://github.com/scylladb/scylla-dtest/pull/4395

**Comment by bhalevy** [comment]
ping @elcallio, is this PR ready to be queued?

**Comment by elcallio** [comment]
Test PR is still no merged, apparently it failed CI on a spurious error (script unpacking, whatnot)... Rerunning it.

**Comment by mykaul** [comment]
@regevran , @scylladb/scylla-maint - please review for merge.

**Comment by avikivity** [comment]
Do we really need rolling restart? We can make the configuration LiveUpdate.

Do we have a way to reconnect without rolling restart?

**Comment by avikivity** [comment]
I don't see any tests.

**Comment by elcallio** [comment]
We can't rolling restart/modify connectors in scylla at this time. 
Test in dtest suite. See link above.

**Comment by avikivity** [comment]
> We can't rolling restart/modify connectors in scylla at this time. Test in dtest suite. See link above.



> We can't rolling restart/modify connectors in scylla at this time. Test in dtest suite. See link above.

We're trying to reduce the dependency on dtest.

File: message/messaging_service.hh
Comment (avikivity) [inline message/messaging_service.hh:263]: Seems to me that transitional is orthogonal to none/rack/dc/all and so should be a separate config item.

encrypt_what: which connections should be encrypted
transitional: whether we ignore mismatches in encrypt_what

Comment (elcallio) [reply message/messaging_service.hh:263]: Not true. Transitional also forces all outgoing connections to be encrypted. I.e. we encrypt outgoint, but accept unencrypted.

**Comment by elcallio** [comment]
> We're trying to reduce the dependency on dtest.

At the time of this PR writing, this policy was not declared. Also, afaik, we don't have infrastructure for cert gen etc in unit tests yet.

File: message/messaging_service.hh
Comment (avikivity) [reply message/messaging_service.hh:263]: It does, but should it? It makes the transition look like none -> all -> dc. Why? Shouldn't we transition none->dc?

Comment (elcallio) [reply message/messaging_service.hh:263]: 1.) dc and rack should be deprecated. I know you want to avoid TLS overhead for local networks etc, but the recent issues with topology not propagating properly and breaking TLS-sct should be enough of an argument that these setups are both brittle as well as potentially insecure. 
2.) We cannot go none -> dc (or rack) without downtime (i.e. restarting all nodes at the same time) since they otherwise will disagree on accepted communication mode. By going this path, we ensure we can do it node-by-node.

**Comment by elcallio** [comment]
Created test in python unit test suite instead/complementing dtest.

File: message/messaging_service.hh
Comment (avikivity) [reply message/messaging_service.hh:271]: I don't see why dc/rack TLS control should be deprecated. dc/rack are stable concepts and it's not like a node can change its dc/rack association at will.

Comment (elcallio) [reply message/messaging_service.hh:271]: Not dc and rack as concepts, but encryption limited to them. Mainly because it is a chicken-egg case, since we need to do interop to know rack/dc relation, and we've seen issues with it breaking when this is not propagated properly. 
But allowing some network traffic to be unencrypted is also a potential security hole if a user fuddles his private network setup in some way.

Comment (avikivity) [reply message/messaging_service.hh:271]: We allow non-TLS connections.

I think it's reasonable to allow connections within the datacenter to be unencrypted and only encrypt them across datacenters.

Didn't we have a mechanism where the initial connection was encrypted, and afterwards we used unencrypted connections if we were in the same rack/dc? Or maybe I'm mixing it with compression.

Comment (elcallio) [reply message/messaging_service.hh:271]: No, we do. But this only affects outgoing connection. If a node A thinks it is known to another node B, and uses non-TLS, assuming the rules should allow this, but the info known in the B target does not match A:s view of itself, we will fail. Which is apparently what happened in the raft-tls issue. 

While I agree about the in-datacenter stuff in principle, my objections are based on a.) the fail(s) like above, and b.) people not securing their "data center" properly - i.e. nw. breaches.

Comment (avikivity) [reply message/messaging_service.hh:271]: If someone chooses not to use TLS, that's on them. I don't think we should enforce it.

Comment (elcallio) [reply message/messaging_service.hh:271]: "enforce" here meaning "apply dc/rack-like rules". In which case we _do_ (and must) enforce TLS iff node (not) belongs to relevant topology group.

File: docs/operating-scylla/security/node-node-encryption.rst
Comment (elcallio) [reply docs/operating-scylla/security/node-node-encryption.rst]: done

**Comment by elcallio** [comment]
@avikivity - final boo or shoo?
@xemul - maintainer, feel free to consider merging.

**Pull request merged**
